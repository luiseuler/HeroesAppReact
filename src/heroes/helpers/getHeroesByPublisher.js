import { heroes } from "../data/heroes"

export const getHeroesByPublisher = (publisher) => {
    const validPublisher = ['DC Comics', 'Marvel Comics'];

    if (!validPublisher.includes(publisher)) {
        throw new Error('El publisher no existe');
    }

    return heroes.filter(hero => hero.publisher === publisher);
}