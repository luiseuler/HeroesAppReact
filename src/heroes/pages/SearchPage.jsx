import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks";
import { getHeroesByName } from "../helpers/getHerosByName";
import { HeroCard } from "../components/HeroCard";

export const SearchPage = () => {
    const navigate = useNavigate();

    const location = useLocation();
    const query = location.search.slice(3);

    const { searchText, onInputChange } = useForm({
        searchText: query
    });


    const heroes = getHeroesByName(query);

    const showSearch = query.length === 0;
    const showError = query.length > 0 && heroes.length === 0;

    const onSearchSubmit = (event) => {
        event.preventDefault();
        navigate(`?q=${searchText}`);
    }

    return (
        <>
            <h1>Search</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Searching</h4>

                    <form aria-label="form"
                        onSubmit={onSearchSubmit}>
                        <input type="text"
                            placeholder="Search a hero"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={searchText}
                            onInput={onInputChange} />

                        <button className="btn btn-outline-primary mt-1">
                            Search
                        </button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr />

                    <div className="alert alert-primary animate__animated animate_fadeIn"
                        style={{display: showSearch ? '' : 'none'}}>
                        Search a Hero
                    </div>

                    <div aria-label="no-hero-label"
                        className="alert alert-danger animate__animated animate_fadeIn"
                    style={{display: showError ? '' : 'none'}}>
                        No hero with <b>{query}</b>
                    </div>

                    {
                        heroes.map(hero => (
                            <HeroCard key={hero.id} {...hero} />
                        ))
                    }
                </div>
            </div>
        </>
    );
}
