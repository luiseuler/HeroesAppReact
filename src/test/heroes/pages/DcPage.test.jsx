import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { DcPage } from "../../../heroes/pages/";
import { getHeroesByPublisher } from "../../../heroes/helpers";

jest.mock('../../../heroes/helpers');

describe('Pruebas en <DcPage />', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('debe de hacer match con el snapshot', () => {
        getHeroesByPublisher.mockReturnValue([
            {
                id: 'dc-batman',
                superhero: 'Batman',
                publisher: 'DC Comics',
                alter_ego: 'Bruce Wayne',
                first_appearance: 'Detective Comics #27',
                characters: 'Bruce Wayne'
            }
        ]);

        const { container } = render(
            <MemoryRouter>
                <DcPage />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });

    test('debe de mostrar el titulo DcPage', () => {
        getHeroesByPublisher.mockReturnValue([
            {
                id: 'dc-batman',
                superhero: 'Batman',
                publisher: 'DC Comics',
                alter_ego: 'Bruce Wayne',
                first_appearance: 'Detective Comics #27',
                characters: 'Bruce Wayne'
            }
        ]);

        render(
            <MemoryRouter>
                <DcPage />
            </MemoryRouter>
        );

        const titleElement = screen.getByRole('heading', {level: 1}).innerHTML;
        
        expect(titleElement).toContain('DcPage');
    });
});