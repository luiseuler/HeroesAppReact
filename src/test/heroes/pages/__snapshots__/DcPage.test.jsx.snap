// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <DcPage /> debe de hacer match con el snapshot 1`] = `
<div>
  <h1>
    DcPage
  </h1>
  <hr />
  <div
    class="row row-cols-1 row-cols-md-3 g-3"
  >
    <div
      class="col animate__animated animate__fadeIn"
    >
      <div
        class="card"
      >
        <div
          class="row no-gutters"
        >
          <div
            class="col-4"
          >
            <img
              alt="Batman"
              class="card-img"
              src="/heroes/dc-batman.jpg"
            />
          </div>
          <div
            class="col-8"
          >
            <div
              class="card-body"
            >
              <h5
                class="card-title"
              >
                Batman
              </h5>
              <p
                class="card-text"
              >
                Bruce Wayne
              </p>
              <p
                class="card-text"
              >
                <small
                  class="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
